Sources/TABResourceLoader/Services/NetworkResponseHandler.swift

```diff
@@ -8,31 +8,82 @@
+///
+/// ---
+/// # Failure case
+///
+/// Defines a response that failed, contains:
+///   - The Result of parsing the model, success case contains the model and failure case contains the parsing error
+///   - A HTTPURLResponse if it exists
+///   - A NetworkServiceError
+///
+/// ## Reasons for failure can be (in order):
+///   1. No HTTPURLResponse with a URLSession error
+///   2. No HTTPURLResponse without a URLSession error (This can only be possible for non HTTP responses)
+///   3. The HTTP status code is 4xx or 5xx
+///   4. A URLSession error exists
+///   5. Could not parse the data
+public enum NetworkResponse<Model> {
+  case success(Model, HTTPURLResponse)
+  case failure(Result<Model>, HTTPURLResponse?, NetworkServiceError)
+}
+
+enum NetworkResponseHandlerError: Error {
+  case noDataProvided
+}
+
 struct NetworkResponseHandler {
 
-  static func resultFrom<Resource: NetworkResourceType & DataResourceType>(resource: Resource, data: Data?, URLResponse: Foundation.URLResponse?, error: Error?) -> Result<Resource.Model> {
+  static func resultFrom<Resource: NetworkResourceType & DataResourceType>(resource: Resource, data: Data?, URLResponse: Foundation.URLResponse?, error: Error?) -> NetworkResponse<Resource.Model> {
+
+    let parsedResult = transform(data, using: resource)
 
-    if let HTTPURLResponse = URLResponse as? HTTPURLResponse {
-      switch HTTPURLResponse.statusCode {
-      case 400..<600:
-        return .failure(NetworkServiceError.statusCodeError(statusCode: HTTPURLResponse.statusCode))
-      default: break
+    guard let HTTPURLResponse = URLResponse as? HTTPURLResponse else {
+      if let error = error {
+        return .failure(parsedResult, nil, .sessionError(error: error))
+      } else {
+        return .failure(parsedResult, nil, .noHTTPURLResponse)
       }
     }
 
+    switch HTTPURLResponse.statusCode {
+    case 400..<600:
+      return .failure(parsedResult, HTTPURLResponse, .statusCodeError(statusCode: HTTPURLResponse.statusCode))
+    default: break
+    }
+
     if let error = error {
-      return .failure(NetworkServiceError.networkingError(error: error))
+      return .failure(parsedResult, HTTPURLResponse, .sessionError(error: error))
     }
 
-    guard let data = data else {
-      return .failure(NetworkServiceError.noData)
+    switch parsedResult {
+    case .success(let model):
+      return .success(model, HTTPURLResponse)
+    case .failure(let parsingError):
+       return .failure(parsedResult, HTTPURLResponse, .couldNotParseData(error: parsingError))
     }
+  }
 
-    do {
-      let parsedResult = try resource.result(from: data)
-      return .success(parsedResult)
-    } catch {
-      return .failure(NetworkServiceError.networkingError(error: error))
+  private static func transform<Resource: DataResourceType>(_ data: Data?, using resource: Resource) -> Result<Resource.Model> {
+    guard let data = data else {
+      return .failure(NetworkResponseHandlerError.noDataProvided)
```

Comment:
my understanding is that it's not possible, you would get `Data()` but not nil, I've never seen this scenario but may be worth exploring what we could do should this be possible